src/metrics/include/metrics/v2/event.hpp

```diff
@@ -0,0 +1,282 @@
+        return str.str();
+    }
+
+    static poplar::CreateOptions createOptions(const std::string& name,
+                                               const std::string& path_prefix) {
+        poplar::CreateOptions options;
+        options.set_name(name);
+        options.set_events(poplar::CreateOptions_EventsCollectorType_BASIC);
+        options.set_path(createPath(name, path_prefix));
+        options.set_chunksize(1000);
+        options.set_streaming(true);
+        options.set_dynamic(false);
+        options.set_recorder(poplar::CreateOptions_RecorderType_PERF);
+        options.set_events(poplar::CreateOptions_EventsCollectorType_BASIC);
+        return options;
+    }
+
+    std::string _name;
+    poplar::PoplarID _id;
+    CollectorStubInterface _stub;
+};
+
+/**
+ * Primary point of interaction between v2 poplar internals and the metrics system.
+ */
+template <typename ClockSource, typename StreamInterface>
+class EventStream {
+    using duration = typename ClockSource::duration;
+    using OptionalPhaseNumber = std::optional<genny::PhaseNumber>;
+
+public:
+    explicit EventStream(const ActorId& actorId,
+                         const std::string& name,
+                         const OptionalPhaseNumber& phase,
+                         const std::string& path_prefix)
+        : _name{name}, _phase{phase}, _last_finish{ClockSource::now()} {
+        _metrics.set_name(_name);
+        _metrics.set_id(actorId);
+        this->_reset();
+    }
+
+    void addAt(const typename ClockSource::time_point& finish,
+               const OperationEventT<ClockSource>& event,
+               size_t workerCount) {
+        _metrics.mutable_time()->set_seconds(
+            std::chrono::time_point_cast<std::chrono::seconds>(finish).time_since_epoch().count());
+        _metrics.mutable_time()->set_nanos(
+            std::chrono::time_point_cast<std::chrono::nanoseconds>(finish)
+                .time_since_epoch()
+                .count());
+
+        _metrics.mutable_timers()->mutable_duration()->set_seconds(
+            event.duration.getSecondsCount());
+        _metrics.mutable_timers()->mutable_duration()->set_nanos(
+            event.duration.getNanosecondsCount());
+
+        _metrics.mutable_timers()->mutable_total()->set_seconds(
+            Period<ClockSource>(finish - _last_finish).getSecondsCount());
+        _metrics.mutable_timers()->mutable_total()->set_nanos(
+            Period<ClockSource>(finish - _last_finish).getNanosecondsCount());
+
+        _metrics.mutable_counters()->set_number(event.iters);
+        _metrics.mutable_counters()->set_ops(event.ops);
+        _metrics.mutable_counters()->set_size(event.size);
+        _metrics.mutable_counters()->set_errors(event.errors);
+
+        _metrics.mutable_gauges()->set_failed(event.isFailure());
+        _metrics.mutable_gauges()->set_workers(workerCount);
+        if (_phase) {
+            _metrics.mutable_gauges()->set_state(*_phase);
+        }
```

Comment:
comment here that "state" is an arbitrary value per the poplar spec and we're using it to store phase.